src/test/osd/RadosModel.h

```diff
@@ -1948,6 +1948,103 @@ class CopyFromOp : public TestOp {
   }
 };
 
+class SetRedirectOp : public TestOp {
+public:
+  string oid, oid_tgt, tgt_pool_name;
+  ObjectDesc src_value;
+  librados::ObjectWriteOperation op;
+  librados::AioCompletion *comp;
+  ceph::shared_ptr<int> in_use;
+  int snap;
+  int done;
+  int r;
+  SetRedirectOp(int n,
+	     RadosTestContext *context,
+	     const string &oid,
+	     const string &oid_tgt,
+	     const string &tgt_pool_name,
+	     TestOpStat *stat)
+    : TestOp(n, context, stat),
+      oid(oid), oid_tgt(oid_tgt), tgt_pool_name(tgt_pool_name),
+      comp(NULL), snap(-1), done(0), 
+      r(0)
+  {}
+
+  void _begin() override
+  {
+    {
+      Mutex::Locker l(context->state_lock);
+      context->oid_in_use.insert(oid);
+      context->oid_not_in_use.erase(oid);
+      context->oid_in_use.insert(oid_tgt);
+      context->oid_not_in_use.erase(oid_tgt);
+
+      // choose source snap
+      if (0 && !(rand() % 4) && !context->snaps.empty()) {
+	snap = rand_choose(context->snaps)->first;
+	in_use = context->snaps_in_use.lookup_or_create(snap, snap);
+      } else {
+	snap = -1;
+      }
+      context->find_object(oid, &src_value, snap);
+      if (!src_value.deleted())
+	context->update_object_full(oid, src_value);
+    }
+
+    librados::IoCtx tgt_ctx;
+    context->rados.ioctx_create(oid_tgt.c_str(), tgt_ctx);
+    op.assert_version(src_value.version);
+    op.set_redirect(oid, tgt_ctx, 0);
```

Comment:
This is incomplete.. there's no change to actually do this with some probability.  Missing a patch maybe?

In any case, I suggest start with the simpler case.  The thing with the redirect is if the target object content changes the src content object will change too, and RadosModel doesn't have anything to handle that sort of thing.  Instead, I think we should do something simpler: in the setup phase of the test, where we're writing the initial N objects, with some probability write the initial object to with a '-redirect-target-nnn suffix (or something unique) and make the real object a redirect to it.  Then no other changes.  That way we are doing all of the same operations we were doing before, with except that it would be proxied through a redirect.

That is much simpler (only need  to modify the setup phase; no need for a new op like this one).

Then, once that is working, we can add SetRedirect and UnsetRedirect (or whatever) operations that do the promote/demote type operations.  This part will be a bit trickier though since we want to make it race with other operations on the same object in order to exercise the race cases.  Let's make the simple one work first!

And put it in a separate commit from the unit test?

Thanks!